### PR TITLE
feat: add move_project_data management command

### PR DESCRIPTION
**Pull request by mihow**
## Summary

Management command to move deployments and all associated data from one project to another. Motivated by a request to move northern Quebec stations (Umiujaq, Kuujjuaq, Kangiqsujuaq, Inukjuak) from "Insectarium de Montreal" into a new "Nunavik" project.

This is a comprehensive data transfer that handles 16 categories of related data:
- Deployments, Events, SourceImages, Occurrences, Jobs (direct project FK updates)
- Detections, Classifications, Identifications (indirect, follow via FK chains)
- Shared resources: S3StorageSources, Devices, Sites (clone if shared, reassign if exclusive)
- Mixed SourceImageCollections (split into source/target copies)
- Pipeline configs and ProcessingService links
- Taxa M2M links for occurrence determinations
- Identifier users added to target project with their existing role preserved
- TaxaLists linked to target project
- Default filter config copied (score threshold, include/exclude taxa, default pipeline)
- Cached field recalculation on deployments, events, and both projects

### Usage

```bash
# Dry run (default) — shows full before/after analysis
python manage.py move_project_data --source-project 23 --create-project "Nunavik" --deployment-ids 84,163,284

# Execute
python manage.py move_project_data --source-project 23 --create-project "Nunavik" --deployment-ids 84,163,284 --execute
```

### Features

- **Dry run by default** with full scope analysis before any changes
- **Conditional scope warning** showing data weight (lightweight for unprocessed images, detailed for processed data with identifications)
- **Per-deployment before/after snapshots** with row counts for all model types
- **Conservation checks** verifying source + target = original totals
- **Validation suite** covering FK integrity, indirect access consistency, leak detection, and collection integrity
- **Atomic transaction** — rolls back cleanly on any failure (validation runs inside the transaction)

### Fixes from review

- Mutual exclusion check for `--target-project` / `--create-project`
- Pipeline config clone preserves `enabled` and `config` fields
- Collection clone preserves `method` and `kwargs` fields
- Validation queries scoped to moved deployments (not entire target project)
- Validation failure now raises `CommandError` (rolls back transaction)
- Project creation uses `Project.objects.create(create_defaults=True)` via ProjectManager
- Bulk taxa linking via `target_project.taxa.add(*taxa_ids)`
- Removed unused `Taxon` import and f-prefix on strings without placeholders

### Takeaway-review fixes (2026-05-01)

- **Job→Collection cross-link fixed.** When a `SourceImageCollection` was mixed (referenced by both moved and staying deployments), the script cloned it to the target project but moved Jobs still pointed at the source collection. Now: `collection_clone_map` is built during the split step and moved Jobs are remapped to the cloned target collection. With `--no-clone-collections`, the FK is nulled instead so no cross-link survives.
- **Validation now fails+rolls back** if any moved Job ends up cross-linked to a SourceImageCollection on a non-target project.
- **`session_time_gap_seconds`** is now copied to the target project (was the one `ProjectSettingsMixin` field missing from the default-filter copy block).
- **`--fire-post-save` flag** added. Off by default. Bulk `.update()` bypasses `post_save` on Event/SourceImage/Occurrence/Job; no listeners exist on those models today, but the flag is the opt-in path if any are added later. Calls full `Deployment.save()` post-commit (also queues Celery event regrouping).
- **Operator notes printed after MOVE COMPLETE**: cache-lag warning, signal status, DataExport residue count, source-side residue (cloned shared resources, multi-project TaxaLists, taxa M2M).
- **`clone_or_reassign` helper** collapses 3× S3/Device/Site loops.
- **Step numbers replaced with semantic labels** (`[create-target]`, `[move-deployment]`, ...) — old `[N/12]` and `[N/16]` ranges were both used and 5 different ops shared `[10/12]`.

### Automated tests (41 tests)

Test file: `ami/main/tests/test_move_project_data.py`

```bash
docker compose -f docker-compose.ci.yml run --rm django python manage.py test ami.main.tests.test_move_project_data --keepdb -v2
```

| Test class | Count | Coverage |
|---|---|---|
| `TestMoveToExistingProject` | 4 | Basic move, taxa linking, conservation counts, jobs moved |
| `TestDryRun` | 1 | No DB mutations without `--execute` |
| `TestCreateProject` | 2 | `--create-project` flag, member copying |
| `TestErrorHandling` | 8 | All input validation error paths (incl. source==target, duplicate deployment ids) |
| `TestSharedResourceCloning` | 7 | Device/Site/S3 clone-vs-reassign, external device unchanged |
| `TestCollectionHandling` | 5 | Exclusive reassign, mixed split (preserves method/kwargs), `--no-clone-collections`, mixed-split remaps moved Jobs, `--no-clone-collections` nulls moved Job collection FK |
| `TestPipelineConfigCloning` | 3 | Config clone with fields, existing not duplicated, `--no-clone-pipelines` |
| `TestProcessingServiceLinking` | 2 | Services linked via raw SQL, no duplicates |
| `TestIdentifierRolePreservation` | 2 | Identifiers added to target, already-member not duplicated |
| `TestTaxaListLinking` | 1 | TaxaLists linked to target project |
| `TestDefaultFilterCopying` | 2 | Score threshold, include/exclude taxa copied |
| `TestEdgeCases` | 4 | Empty deployment, move all, multiple deployments, pre-existing taxa |

### Manual validation (pre-production)

Before running on production data, validate on a staging DB or local backup:

1. **Dry run first** — review the full output, check per-deployment counts match expectations
   ```bash
   python manage.py move_project_data --source-project 23 --create-project "Nunavik" --deployment-ids 84,163,284
   ```

2. **Execute** — review the validation output at the end
   ```bash
   python manage.py move_project_data --source-project 23 --create-project "Nunavik" --deployment-ids 84,163,284 --execute
   ```

3. **Post-move spot checks** in Django admin or shell:
   - Source project's deployment list no longer shows moved deployments
   - Target project's deployments, events, and occurrences are browsable
   - Images load correctly in the UI (S3 paths are unchanged)
   - Processing pipelines are available on the target project
   - Occurrence filters and taxa lists work on the target
   - Identifier users can log in and see their identifications in the target project

4. **If validation fails**, the transaction is rolled back automatically and a `CommandError` is raised with details. No manual cleanup needed.

## Test plan

- [x] 41 automated tests passing
- [x] Tested locally on BCI project (2 deployments, 73k images, 57k occurrences, 81k classifications, 751 identifications)
- [x] All per-deployment row counts preserved after move
- [x] Conservation checks passed (source + target = original)
- [x] FK integrity verified (all moved data points to target project)
- [x] Indirect access verified (detections, classifications, identifications resolve correctly via project)
- [x] No data leaked to source project
- [x] Mixed SourceImageCollections split correctly
- [x] Shared devices cloned, NULL-project devices left as-is
- [ ] Test on staging with real Nunavik data before production use

## Summary by CodeRabbit

* **New Features**
  * New command to move one or more deployments between projects with optional atomic execution
  * Dry-run preview mode and option to create the target project
  * Options for cloning or reassigning shared resources, splitting mixed image collections, copying pipeline configs, linking processing services, and preserving identifier roles and default filters
  * Comprehensive in-transaction validation and post-move recalculation

* **Documentation**
  * Added deployment reassignment guide and project portability specification

* **Tests**
  * Extensive end-to-end test coverage for move scenarios and edge cases

### CODE REVIEW

File: docs/claude/planning/deployment-reassignment-guide.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+| **Classification** | `detection.source_image.project` | Follows Detection chain |
+| **Identification** | `occurrence.project` | Follows Occurrence automatically |
+
+### Tier 3 — Shared/linked resources (may need cloning or re-linking)
+
+| Resource | Relationship | Reassignment strategy |
+|---|---|---|
+| **S3StorageSource** | FK on Deployment (`data_source`) | Clone if `project_id` points to source project; update deployment FK |
+| **Device** | FK on Deployment | Clone if `project_id` = source project; or set NULL (shared) |
+| **Site** | FK on Deployment (`research_site`) | Clone if `project_id` = source project; or set NULL (shared) |
```
[the remark this answers is not in the document]
Comment (mihow) [reply docs/claude/planning/deployment-reassignment-guide.md]: **Claude says:** Valid. The guide was written before the shared-vs-exclusive logic was finalized. Will update in a follow-up.

File: ami/main/management/commands/move_project_data.py [hunk 2]
```diff
@@ -0,0 +1,766 @@
...
+
+        # Update both projects' related calculated fields (events + deployments)
+        self.log(f"  Updating source project cached fields...")
+        source_project.update_related_calculated_fields()
+        self.log(f"  Source project '{source_project.name}': related fields updated")
+
+        self.log(f"  Updating target project cached fields...")
+        target_project.update_related_calculated_fields()
+        self.log(f"  Target project (id={target_id}): related fields updated")
+
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Valid perf concern. For this admin-only command (used on a handful of projects), the N+1 is acceptable. Cached field updates are now outside the transaction (idempotent, re-runnable). A bulk recalculation helper would be a good future optimization.

File: ami/main/management/commands/move_project_data.py [hunk 3]
```diff
@@ -0,0 +1,766 @@
...
+            if include_taxa or exclude_taxa:
+                for t in source_project.default_filters_include_taxa.all():
+                    target_project.default_filters_include_taxa.add(t)
+                for t in source_project.default_filters_exclude_taxa.all():
+                    target_project.default_filters_exclude_taxa.add(t)
+                self.log("  [15/16] Copied default filter taxa config")
+            target_project.default_filters_score_threshold = source_project.default_filters_score_threshold
+            if source_project.default_processing_pipeline:
+                target_project.default_processing_pipeline = source_project.default_processing_pipeline
+            target_project.save()
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Intentional. Additive is safer than replacing — we do not want to silently overwrite an existing target projects filters. For --create-project targets this is identical to replace. Score threshold is always copied (line 602).

File: ami/main/management/commands/move_project_data.py [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+                        description=coll.description or "",
+                    )
+                    new_coll.images.set(moved_image_ids)
+                    self.log(
+                        f"  [10/12] Split collection '{coll.name}': "
+                        f"cloned {len(moved_image_ids):,} images → target collection id={new_coll.pk}"
+                    )
+
+                coll.images.remove(*moved_image_ids)
+                self.log(f"  [10/12] Removed {len(moved_image_ids):,} images from source collection '{coll.name}'")
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Valid for very large collections. For our use case (tens of thousands per collection max), materializing PKs is fine. The M2M .remove() uses a single DELETE WHERE IN query. Not blocking.

File: ami/main/management/commands/move_project_data.py [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+
+            # 10. Clone pipeline configs
+            if clone_pipelines:
+                existing_pipelines = set(
+                    ProjectPipelineConfig.objects.filter(project_id=target_id).values_list("pipeline_id", flat=True)
+                )
+                cloned_count = 0
+                for config in ProjectPipelineConfig.objects.filter(project_id=source_project_id):
+                    if config.pipeline_id not in existing_pipelines:
+                        ProjectPipelineConfig.objects.create(project_id=target_id, pipeline_id=config.pipeline_id)
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Already fixed in ae41cb20. Lines 546-548 now copy `enabled` and `config` fields.

File: ami/main/management/commands/move_project_data.py [hunk 6]
```diff
@@ -0,0 +1,766 @@
...
+        dets_via_dep = Detection.objects.filter(source_image__deployment_id__in=deployment_ids).count()
+        if dets_via_project != dets_via_dep:
+            errors.append(f"Detection count mismatch: via project={dets_via_project}, via deployment={dets_via_dep}")
+            self.log(
+                f"  FAIL: Detection count mismatch: via project={dets_via_project}, via deployment={dets_via_dep}"
+            )
+        else:
+            self.log(
+                f"  OK: Detections consistent ({dets_via_project:,} via project, {dets_via_dep:,} via deployment)"
+            )
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Already fixed in ae41cb20. Validation queries now filter by both project_id AND deployment_id to handle pre-existing target data.

File: ami/main/management/commands/move_project_data.py [hunk 7]
```diff
@@ -0,0 +1,766 @@
...
+                self.log(f"  [11/12] Pipeline configs: cloned {cloned_count}, target now has {total}")
+
+            # 11. Link ProcessingServices (raw SQL to avoid ORM column mismatch)
+            linked = link_processing_services_raw(source_project_id, target_id)
+            self.log(f"  [11/12] Linked {linked} ProcessingService(s) to target project")
+
+            # 12. Link taxa to target project
+            if taxa_ids:
+                for taxon in Taxon.objects.filter(pk__in=taxa_ids):
+                    taxon.projects.add(target_project)
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Already uses bulk add at line 559: `target_project.taxa.add(*taxa_ids)`. Single query.

File: ami/main/management/commands/move_project_data.py [hunk 8]
```diff
@@ -0,0 +1,766 @@
...
+        # === EXECUTE ===
+        self.log(f"\n{'─' * 60}")
+        self.log("  EXECUTING MOVE")
+        self.log(f"{'─' * 60}")
+
+        with transaction.atomic():
+            # 0. Create target project inside transaction
+            if create_project_name:
+                target_project = Project(name=create_project_name, owner=source_project.owner)
+                target_project.save()
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Not applicable. Line 414 uses `Project.objects.create(create_defaults=True)` which goes through the ProjectManager and calls `create_related_defaults()`.

File: ami/main/management/commands/move_project_data.py [hunk 9]
```diff
@@ -0,0 +1,766 @@
...
+            self.log(
+                f"  OK: Detections consistent ({dets_via_project:,} via project, {dets_via_dep:,} via deployment)"
+            )
+
+        cls_via_project = Classification.objects.filter(detection__source_image__project_id=target_id).count()
+        cls_via_dep = Classification.objects.filter(detection__source_image__deployment_id__in=deployment_ids).count()
+        if cls_via_project != cls_via_dep:
+            errors.append(
+                f"Classification count mismatch: via project={cls_via_project}, via deployment={cls_via_dep}"
+            )
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Fixed in ae41cb20. Classification mismatch now has a FAIL log line consistent with the others.

File: ami/main/management/commands/move_project_data.py [hunk 10]
```diff
@@ -0,0 +1,766 @@
...
+
+        # --- Final verdict ---
+        self.log(f"\n{'=' * 60}")
+        if errors:
+            self.log("  VALIDATION FAILED", style=self.style.ERROR)
+            for err in errors:
+                self.log(f"    ✗ {err}", style=self.style.ERROR)
+        else:
+            self.log("  ALL VALIDATION CHECKS PASSED", style=self.style.SUCCESS)
+        self.log(f"{'=' * 60}")
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Fixed in ae41cb20 (raises CommandError). Further improved: validation now runs inside the transaction, so failures trigger a full rollback.

File: ami/main/management/commands/move_project_data.py [hunk 11]
```diff
@@ -0,0 +1,766 @@
...
+                user = User.objects.get(pk=uid)
+                source_role = Role.get_primary_role(source_project, user)
+                if source_role:
+                    role_to_assign = source_role
+                    role_source = "source project role"
+                else:
+                    role_to_assign = Identifier
+                    role_source = "default (not a source member)"
+                identifier_role_map[uid] = role_to_assign
+                self.log(f"    {user.email}: " f"{role_to_assign.display_name} ({role_source})")
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Fixed. Email addresses replaced with `user id=N` in log output.

File: docs/claude/planning/deployment-reassignment-guide.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mihow) [reply docs/claude/planning/deployment-reassignment-guide.md]: **Claude says:** Acknowledged. Guide will be updated to match command behavior in a follow-up.

File: ami/main/management/commands/move_project_data.py [hunk 12]
```diff
@@ -0,0 +1,790 @@
...
+        except Project.DoesNotExist:
+            raise CommandError(f"Source project {source_project_id} does not exist")
+
+        self.log(f"\nSource project: {source_project.name} (id={source_project.pk})")
+
+        deployments = Deployment.objects.filter(pk__in=deployment_ids)
+        if deployments.count() != len(deployment_ids):
+            found = set(deployments.values_list("pk", flat=True))
+            missing = set(deployment_ids) - found
+            raise CommandError(f"Deployments not found: {missing}")
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py:205]: **Claude says:** Fixed. Deployment IDs are now deduplicated via `dict.fromkeys()` at parse time, preserving order. Added test `test_duplicate_deployment_ids` to verify.

File: ami/main/management/commands/move_project_data.py [hunk 13]
```diff
@@ -0,0 +1,790 @@
...
+            target_project = None
+        elif options.get("target_project"):
+            try:
+                target_project = Project.objects.get(pk=options["target_project"])
+            except Project.DoesNotExist:
+                raise CommandError(f"Target project {options['target_project']} does not exist")
+            create_project_name = None
+            self.log(f"Target project: {target_project.name} (id={target_project.pk})")
+        else:
+            raise CommandError("Must specify either --target-project or --create-project")
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py:230]: **Claude says:** Fixed. Added explicit check at line 177 that raises CommandError when target equals source. Added test `test_source_equals_target`.

File: ami/main/management/commands/move_project_data.py [hunk 14]
```diff
@@ -0,0 +1,790 @@
...
+                style=self.style.WARNING,
+            )
+            return
+
+        # === EXECUTE ===
+        self.log(f"\n{'─' * 60}")
+        self.log("  EXECUTING MOVE")
+        self.log(f"{'─' * 60}")
+
+        with transaction.atomic():
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py:457]: **Claude says:** Good catch. Restructured: all validation checks now run inside the `transaction.atomic()` block. If validation fails, `CommandError` is raised inside the transaction, triggering a full rollback. Cached field updates remain outside (idempotent, safe to re-run).

File: ami/main/management/commands/move_project_data.py [hunk 15]
```diff
@@ -0,0 +1,790 @@
...
+            if include_taxa or exclude_taxa:
+                for t in source_project.default_filters_include_taxa.all():
+                    target_project.default_filters_include_taxa.add(t)
+                for t in source_project.default_filters_exclude_taxa.all():
+                    target_project.default_filters_exclude_taxa.add(t)
+                self.log("  [15/16] Copied default filter taxa config")
+            target_project.default_filters_score_threshold = source_project.default_filters_score_threshold
+            if source_project.default_processing_pipeline:
+                target_project.default_processing_pipeline = source_project.default_processing_pipeline
+            target_project.save()
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/main/management/commands/move_project_data.py]: **Claude says:** Intentional design choice. Additive (merge) is safer than replace for existing targets — we should not silently overwrite a target projects existing filter configuration. For --create-project (the primary use case), the target starts empty so additive = replace. Score threshold is always overwritten (line 602).
